### PR TITLE
Fix workflow operators

### LINKED ISSUE #69: switchMapState isn't actually a switchMap, it's a concatMap

### LINKED ISSUE #71: Result of mapResult operator doesn't propagate cancellation upstream
If you do something like `source.mapResult { it }.cancel()`, `source` will not be cancelled.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
* 9c879cd Fix #69 by correctly implementing switchMapState operator.
* e304784 Fix #71 by cancelling upstream workflow when mapResult workflow is cancelled.
* 1876788 Move the workflow operators into their own file and write unit tests.

### CODE REVIEW

File: workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import kotlinx.coroutines.experimental.GlobalScope
+import kotlinx.coroutines.experimental.async
+import kotlinx.coroutines.experimental.channels.Channel.Factory.CONFLATED
+import kotlinx.coroutines.experimental.channels.ReceiveChannel
+import kotlinx.coroutines.experimental.channels.consumeEach
+import kotlinx.coroutines.experimental.channels.produce
+import kotlin.coroutines.experimental.CoroutineContext
+
+/**
+ * [CoroutineContext] used by [Workflow] operators below.
```
Comment (rjrjr) [inline workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt]: Could explain why you chose `Unconfined`. (Follow up for nits is fine w/me.)

File: workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt [hunk 2]
```diff
@@ -73,10 +76,36 @@ fun <S1 : Any, S2 : Any, E : Any, O : Any> Workflow<S1, E, O>.switchMapState(
...
-        transform(rawState).consumeEach { transformedState ->
-          send(transformedState)
+      val upstreamChannel = this@switchMapState.openSubscriptionToState()
+      var transformedChannel: ReceiveChannel<S2>? = null
+      val downstreamChannel = channel
+
+      upstreamChannel.consume {
+        whileSelect {
+          upstreamChannel.onReceiveOrNull { upstreamState ->
+            if (upstreamState == null) {
```
Comment (rjrjr) [inline workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt]: I wonder if both of these would be more readable flipped and converted to use `?.let / ?:`.

Comment (zach-klippenstein) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt]: That feels like code golf to me. `let` improves readability when the nullable value is being transformed, in this case it's not, it's being used as a sentinel for a completely different signal (channel close). This isn't ideal obviously and there's an open issue to make a more explicit receive call that doesn't use null.

File: workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt (hunk 1 above)
Comment (zach-klippenstein) [reply workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt]: https://github.com/square/workflow/issues/74

**Pull request merged**